newIDE/app/src/guidelines/index.js

```diff
@@ -0,0 +1,205 @@
+// @flow
+import React, { PureComponent } from 'react';
+import ReactDOM from 'react-dom';
+import Button from '@material-ui/core/Button';
+import Popper from '@material-ui/core/Popper';
+import Paper from '@material-ui/core/Paper';
+import Typography from '@material-ui/core/Typography';
+import guidelines from './data.js';
+import Window from '../Utils/Window';
+import { getHelpLink } from '../Utils/HelpLink';
+import { Line } from '../UI/Grid';
+
+//TODO wrap l'image dans le Paper avec l'overflow hidden
+//Actuellement l'image dépasse du Paper alors qu'il y a une marge autour du Popper
+//Actualisé la position de la Popper.
+
+const styles = {
+  container: {
+    maxWidth: 400,
+    padding: 10,
+    marginTop: 10,
+    overflow: 'hidden',
+  },
+  description: {
+    paddingBottom: 10,
+  },
+};
+
+type State = {|
+  open: boolean,
+  index: number,
+  indexData: number,
+  indexMax: number,
+  anchor: *,
+  arrowRef: *,
+|};
+
+type Props = {|
+  open: boolean,
+  closeHandler: () => void,
+|};
+
+//TODO
+//PureComponent ou React.Component<Props, State> voir la doc pour comprendre la différence.
+//Le bouton restart du tuto ferme le Popper mais le reouvre pas et surtout ne reset pas le Popper a l'index 0...
+
+export default class GuidelinePopOver extends PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      open: props.open,
+      index: 0,
+      indexData: 0,
+      indexMax: guidelines.length - 1,
+      anchor: null,
+      arrowRef: null,
+    };
+  }
+
+  componentWillReceiveProps(newProps: *) {
+    let elementHTML = document.querySelectorAll(
+      '.guideline-' + guidelines[this.state.indexData].positionBind
+    )[0];
+    const element = ReactDOM.findDOMNode(elementHTML);
+
+    if (newProps.open !== this.props.open) {
+      this.setState({
+        open: newProps.open,
+        anchor: element,
+      });
+    }
+  }
+
+  next = () => {
+    this.setState(currentState => ({
+      index: Math.min(currentState.indexMax, currentState.index + 1),
+    }));
+  };
+
+  back = () => {
+    this.setState(currentState => ({
+      index: Math.min(currentState.indexMax, currentState.index - 1),
+    }));
+  };
+
+  componentDidMount() {}
+
+  componentDidUpdate(prevProps: *, prevState: *) {
+    if (this.state.index !== prevState.index) {
+      this.setState({ indexData: this.state.index });
```

Comment:
Duplicating a variable of a state is in 99% of the cases, a sign that things are too complicated. Instead, you want to update the position of the `Popper` when the index is changed. So you do:

```
if (this.state.index !== prevState.index) {
  this. _updatePopover();
}
```

instead, with `_updatePopover` being the logic calling `querySelectorAll` and updating `anchor`. As `_updatePopover` will update the state `anchor`, the component will be re-rendered with the proper `anchor` passed as a prop to the Popper.

And you actually don't even need `componentDidUpdate`. Instead, call `this._updatePopover()` whenever you change `index`. This means that indeed of doing:

1. You change `index` state in some function
2. React re-renders the Guidelines component
3. React calls Guidelines `componentDidUpdate`
4. You see that index changed, so you call `_updatePopover`
5. `_updatePopover` updates the `anchor` state.
6. React re-renders the Guidelines component
7. Finally, Popper is re-rendered with the proper Anchor

you just do:

1. You change `index` state in some function
2. You call `_updatePopover`, which updates the `anchor` state.
3. React re-renders the Guidelines component
4. Popper is re-rendered with the proper Anchor

which is a bit "simpler" :)

Reply:
To do something just after the state is updated, use a second argument for setState:

```
this.setState( index: ..., () => this._updatePopover())
```

This guarantees that `this.state.index` will have the new value when `this._updatePopover()` is called. (because React can batch the calls to setState, so you risk to read a value of this.state.index that is NOT updated if you call `_updatePopover` just after `this.setState`.)

Reply:
Wow really better like this ! Refresh working now 💯 
(A thing, callback is executed after the execution of the function so in fact we said to React to make a rendering throught setState(), then it starts the callback but our callback calls a setState() which restarts a rendering.)

Reply:
Yup strictly speaking as we call updatePopover in a callback there are two renderings. It's not a big problem, you could maybe avoid it entirely by passing an index argument to updatePopover and doing only one setState.

But it's not a big deal, the most important is to keep a state that is clean and simple :) (i.e: no extra variable, no calls to the "weird" React lifecycle methods if we can avoid)